fakestorage/server.go

```diff
@@ -259,10 +259,13 @@ func (s *Server) buildMuxer() {
 
 	for _, r := range routers {
 		r.Path("/b").Methods(http.MethodGet).HandlerFunc(jsonToHTTPHandler(s.listBuckets))
+		r.Path("/b/").Methods(http.MethodGet).HandlerFunc(jsonToHTTPHandler(s.listBuckets))
 		r.Path("/b").Methods(http.MethodPost).HandlerFunc(jsonToHTTPHandler(s.createBucketByPost))
+		r.Path("/b/").Methods(http.MethodPost).HandlerFunc(jsonToHTTPHandler(s.createBucketByPost))
 		r.Path("/b/{bucketName}").Methods(http.MethodGet).HandlerFunc(jsonToHTTPHandler(s.getBucket))
 		r.Path("/b/{bucketName}").Methods(http.MethodDelete).HandlerFunc(jsonToHTTPHandler(s.deleteBucket))
 		r.Path("/b/{bucketName}/o").Methods(http.MethodGet).HandlerFunc(jsonToHTTPHandler(s.listObjects))
+		r.Path("/b/{bucketName}/o/").Methods(http.MethodGet).HandlerFunc(jsonToHTTPHandler(s.listObjects))
 		r.Path("/b/{bucketName}/o").Methods(http.MethodPost).HandlerFunc(jsonToHTTPHandler(s.insertObject))
```

Comment:
Note that I'm ignoring this route on purpose. I'm assuming it will be deleted in https://github.com/fsouza/fake-gcs-server/pull/1251.

Reply:
What's the behavior of GCS in this case? Does it redirect or does it just work? Can you provide an example?

Reply:
It just works. The paths with the trailing slash behave exactly like the ones without the trailing slash. There is no redirection involved.

The examples below show the HTTP response (omitting the body and redacting a couple of header values) with and without trailing slash when requesting a list of buckets or uploading an object. There is no change between absence and presence of trailing slash.

```shell
# Listing GCS buckets in "my-project" (without trailing "/")
$ curl -v -s 'https://storage.googleapis.com/storage/v1/b?project=my-project' -H "Authorization: Bearer my-token" 2>&1 | grep '^<' | awk '/uploadid/ {$3 = "REDACTED"} {print}'
< HTTP/2 200 
< x-guploader-uploadid: REDACTED
< content-type: application/json; charset=UTF-8
< date: Sat, 29 Jul 2023 13:04:32 GMT
< vary: Origin
< vary: X-Origin
< cache-control: private, max-age=0, must-revalidate, no-transform
< expires: Sat, 29 Jul 2023 13:04:32 GMT
< content-length: 32
< server: UploadServer
< alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
< 
```

```shell
# Listing GCS buckets in "my-project" (with trailing "/")
$ curl -v -s 'https://storage.googleapis.com/storage/v1/b/?project=my-project' -H "Authorization: Bearer my-token" 2>&1 | grep '^<' | awk '/uploadid/ {$3 = "REDACTED"} {print}'
< HTTP/2 200
< x-guploader-uploadid: REDACTED
< content-type: application/json; charset=UTF-8
< date: Sat, 29 Jul 2023 13:04:36 GMT
< vary: Origin
< vary: X-Origin
< cache-control: private, max-age=0, must-revalidate, no-transform
< expires: Sat, 29 Jul 2023 13:04:36 GMT
< content-length: 32
< server: UploadServer
< alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
< 
```

```shell
# Uploading "Some content" to "gs://my-bucket/my-file" (without trailing "/")
$ curl -v -s -X POST 'https://storage.googleapis.com/upload/storage/v1/b/my-bucket/o?uploadType=media&name=my-file' -d 'Some content' -H "Authorization: Bearer my-token" 2>&1 | grep '^<' | awk '/uploadid/ || /etag/ {$3 = "REDACTED"} {print}'
< HTTP/2 200 
< x-guploader-uploadid: REDACTED
< etag: REDACTED
< content-type: application/json; charset=UTF-8
< date: Sat, 29 Jul 2023 13:00:16 GMT
< vary: Origin
< vary: X-Origin
< cache-control: no-cache, no-store, max-age=0, must-revalidate
< expires: Mon, 01 Jan 1990 00:00:00 GMT
< pragma: no-cache
< content-length: 856
< server: UploadServer
< alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
< 
```

```shell
# Uploading "Some content" to "gs://my-bucket/my-file" (with trailing "/")
$ curl -v -s -X POST 'https://storage.googleapis.com/upload/storage/v1/b/my-bucket/o/?uploadType=media&name=my-file' -d 'Some content' -H "Authorization: Bearer my-token" 2>&1 | grep '^<' | awk '/uploadid/ || /etag/ {$3 = "REDACTED"} {print}'
< HTTP/2 200 
< x-guploader-uploadid: REDACTED
< etag: REDACTED
< content-type: application/json; charset=UTF-8
< date: Sat, 29 Jul 2023 13:00:18 GMT
< vary: Origin
< vary: X-Origin
< cache-control: no-cache, no-store, max-age=0, must-revalidate
< expires: Mon, 01 Jan 1990 00:00:00 GMT
< pragma: no-cache
< content-length: 856
< server: UploadServer
< alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
< 
```

When it comes to the path that will be removed in https://github.com/fsouza/fake-gcs-server/pull/1251, it does returns a `location` header in its response. But it is still a 400 BadRequest response and not a redirect; adding `-L` in the `curl` command below still results in a 400 BadRequest.

```shell
# Uploading "Some content" to "gs://my-bucket/my-file" (without /upload)
$ curl -v -s -X POST 'https://storage.googleapis.com/storage/v1/b/my-bucket/o/?uploadType=media&name=my-file' -d 'Some content' -H "Authorization: Bearer my-token" 2>&1 | grep '^<' | awk '/uploadid/ || /etag/ {$3 = "REDACTED"} {print}'
< HTTP/2 400 
< x-guploader-uploadid: REDACTED
< location: https://storage.googleapis.com/upload/storage/v1/b/my-bucket/o?uploadType=media&name=my-file
< date: Sat, 29 Jul 2023 13:12:38 GMT
< vary: Origin
< vary: X-Origin
< cache-control: no-cache, no-store, max-age=0, must-revalidate
< expires: Mon, 01 Jan 1990 00:00:00 GMT
< pragma: no-cache
< content-length: 568
< server: UploadServer
< content-type: text/html; charset=UTF-8
< alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
< 
```

Reply:
Awesome, thanks for confirming!